store/src/main/java/org/apache/rocketmq/store/ha/autoswitch/AutoSwitchHAConnection.java

```diff
@@ -0,0 +1,597 @@
+            AutoSwitchHAConnection.LOGGER.info(this.getServiceName() + " service started");
+
+            while (!this.isStopped()) {
+                try {
+                    this.selector.select(1000);
+
+                    switch (currentState) {
+                        case HANDSHAKE:
+                            // Wait until the slave send it handshake msg to master.
+                            if (!isSlaveSendHandshake) {
+                                this.waitForRunning(10);
+                                continue;
+                            }
+
+                            if (this.lastWriteOver) {
+                                if (!buildHandshakeBuffer()) {
+                                    LOGGER.error("AutoSwitchHAConnection build handshake buffer failed");
+                                    this.waitForRunning(5000);
+                                    continue;
+                                }
+                            }
+
+                            this.lastWriteOver = handshakeWithSlave();
+                            if (this.lastWriteOver) {
+                                // change flag to {false} to wait for slave notification
+                                isSlaveSendHandshake = false;
+                            }
+                            break;
+                        case TRANSFER:
+                            if (-1 == slaveRequestOffset) {
+                                this.waitForRunning(10);
+                                continue;
+                            }
+
+                            if (-1 == this.nextTransferFromWhere) {
+                                if (0 == slaveRequestOffset) {
+                                    this.nextTransferFromWhere = haService.getDefaultMessageStore().getCommitLog().getMinOffset();
+                                } else {
+                                    this.nextTransferFromWhere = slaveRequestOffset;
+                                }
+                                // Setup initial transferEpoch
+                                EpochEntry epochEntry = AutoSwitchHAConnection.this.epochCache.findEpochEntryByOffset(this.nextTransferFromWhere);
+                                if (epochEntry == null) {
+                                    LOGGER.error("Failed to find an epochEntry to match slaveRequestOffset {}", this.nextTransferFromWhere);
+                                    waitForRunning(500);
+                                    break;
+                                }
+                                AutoSwitchHAConnection.this.currentTransferEpoch = epochEntry.getEpoch();
+                                AutoSwitchHAConnection.this.currentTransferEpochEndOffset = epochEntry.getEndOffset();
+                                if (epochEntry.getEpoch() == AutoSwitchHAConnection.this.epochCache.lastEpoch()) {
+                                    AutoSwitchHAConnection.this.currentTransferEpochEndOffset = Long.MAX_VALUE;
```

Comment:
Although it is almost impossible to exceed Long.MAX_VALUE, it would be to change it to - 1 for infinity.